.github/workflows/ci.yml

```diff
@@ -84,32 +84,38 @@ jobs:
         name: Prepare LD_LIBRARY_PATH (Ubuntu only)
         run: echo LD_LIBRARY_PATH=${pythonLocation}/lib >> $GITHUB_ENV
 
-      - name: Build docs
-        run: cargo doc --no-default-features --features "macros num-bigint num-complex hashbrown" --verbose --target ${{ matrix.platform.rust-target }}
+      - name: Prepare workflow settings
+        id: settings
+        shell: bash
+        run: |
+          if [[ $PYPY == "true" ]]
+          then
+            # On PyPy can't execute tests because no embedding API.
+            echo "::set-output name=cargo_command::cargo build --lib --tests --no-default-features --verbose"
+          else
+            echo "::set-output name=cargo_command::cargo test --no-default-features --verbose"
+          fi
+          echo "::set-output name=all_additive_features::macros num-bigint num-complex hashbrown serde"
+        env:
+          PYPY: ${{ matrix.python-version == 'pypy-3.6' }}
 
-      - name: Build (no features)
-        run: cargo build --no-default-features --verbose --target ${{ matrix.platform.rust-target }}
+      - name: Build docs
+        run: cargo doc --no-default-features --features "${{ steps.settings.outputs.all_additive_features }}" --verbose
 
-      - name: Build (all additive features)
-        run: cargo build --no-default-features --features "macros num-bigint num-complex hashbrown serde" --verbose --target ${{ matrix.platform.rust-target }}
+      - name: Test (no features)
+        run: ${{ steps.settings.outputs.cargo_command }}
 
-      # Run tests (except on PyPy, because no embedding API).
-      - if: matrix.python-version != 'pypy-3.6'
-        name: Test
-        run: cargo test --no-default-features --features "macros num-bigint num-complex hashbrown serde" --target ${{ matrix.platform.rust-target }}
+      - name: Test (all additive features)
+        run: ${{ steps.settings.outputs.cargo_command }} --features "${{ steps.settings.outputs.all_additive_features }}"
```

Comment:
How about something like
```yaml
       run: |
          if [[ $PYPY == "true" ]]; then
              cargo build --no-default-features --features "${{ steps.settings.outputs.all_additive_features }}" --target ${{ matrix.platform.rust-target }}
             cargo test --no-default-features --features "abi3 ${{ steps.settings.outputs.all_additive_features }}" --target ${{ matrix.platform.rust-target }}
          else
              cargo test --no-default-features --features "abi3 ${{ steps.settings.outputs.all_additive_features }}" --target ${{ matrix.platform.rust-target }}
              cargo test --no-default-features --features "abi3 ${{ steps.settings.outputs.all_additive_features }}" --target ${{ matrix.platform.rust-target }}
...
          fi
```

Reply:
I like having the different feature combinations as different steps, because it makes it much easier to see which combinations of features went wrong.